devdoc/waiter_queue_requirements.md

```diff
@@ -0,0 +1,103 @@
+# `waiter_queue` requirements
+
+## Overview
+
+`waiter_queue` is a module that allows consumers to park calls and for producers to signal that the parked calls be executed. A "parked call" consists of a callback function and a context. The consumer provides the callback function and context when it parks a call. The consumer expects the parked call's callback function to be executed with the provided context when the producer signals.
+
+`waiter_queue` has two parts: `waiter_queue_ll` that contains the logic for parking and unblocking calls and `waiter_queue_hl` that contains logic for making `waiter_queue` thread-safe.
+
+Consumers call `waiter_queue_hl_add_waiter` with a `UNBLOCK_CALLBACK` callback function and context to park a call. The parked call is added to a queue and will be executed when the producers signals by calling `waiter_queue_ll_unblock_waiters`.
+
+Producers call `waiter_queue_ll_unblock_waiters` with the data they would like to provide to the parked calls. When a producer calls `waiter_queue_ll_unblock_waiters` and there are parked calls in the queue, the parked calls are executed with the given data and `WAITER_QUEUE_CALL_REASON_UNBLOCKED` as `reason`.
+
+Calls are executed in the FIFO order in which they were parked. If the `UNBLOCK_CALLBACK` for a parked call sets the `remove` out argument to `true`, that call is removed from the queue. Otherwise, the parked call stays in the queue and may be executed again when the producer signals again.
+
+A single call to `waiter_queue_ll_unblock_waiters` may result in executing multiple parked calls. `waiter_queue` executes each `UNBLOCK_CALLBACK` in FIFO order until a `UNBLOCK_CALLBACK` sets the `continue_processing` out argument to `false`.
+
+`waiter_queue_hl_destroy` calls `waiter_queue_ll_abandon` to abandon all parked calls. Each parked call that is in the queue when `waiter_queue_hl_destroy` is called will have its `UNBLOCK_CALLBACK` executed with `NULL` as `data` and `WAITER_QUEUE_CALL_REASON_ABONDONED` as `reason`. 
+
+## Example:
+
+`waiter_queue_hl_create` is called.
+
+Queue state: `head||tail`
+
+Consumer calls `waiter_queue_hl_add_waiter(cb1, ctx1)`
+
+Queue state: `head|(cb1, ctx1)|tail`
+
+Consumer calls `waiter_queue_hl_add_waiter(cb2, ctx2)`
+
+Queue state: `head|(cb1, ctx1)|(cb2, ctx2)|tail`
+
+Producer calls `waiter_queue_hl_unblock_waiters(data1)`
+
+`cb1` is called with `ctx1`, `data1` and `WAITER_QUEUE_CALL_REASON_UNBLOCKED`. `cb1` sets `remove` to `true` and `continue_processing` to `true`.
+
+Queue state: `head|(cb2, ctx2)|tail`
+
+`cb2` is called with `ctx2`, `data1` and `WAITER_QUEUE_CALL_REASON_UNBLOCKED`. `cb2` sets `remove` to `false` and `continue_processing` to `true`.
+
+Queue state: `head|(cb2, ctx2)|tail`
+
+Consumer calls `waiter_queue_hl_add_waiter(cb3, ctx3)`.
```

Comment:
I suppose in this example this line happens from a different thread, but if `cb2` actually made the call to `waiter_queue_hl_add_waiter(cb3, ctx3)` then cb3 would execute immediately.

I ask because in the one use case I have, if a waiter is unblocked with data that indicates an error, it will bubble that error up and then the callback handler for that will retry, which would re-queue all on the same thread in the execution context of the waiter callback. And this had infinite loops in my original implementation without a fix to break the cycle.

So I suppose I would still need to handle that unless the `waiter_queue_hl_unblock_waiters` only works off of a snapshot of the queue of waiters when it starts executing.

Reply:
We can implement it such that `waiter_queue_hl_unblock_waiters` causes only the parked calls that were parked before are executed. Adding verbiage for this behavior in the next iteration.